### PR TITLE
Simplify dimensions of

### PR DESCRIPTION
**Pull request by CharlieL7**
* Simplifies `dimensions_of` instructions to a literal when possible.
* Intended to be used after the `split_single_dyn_dim` pass.

### CODE REVIEW

**Review by bpickrel — APPROVED** [review]

File: src/simplify_dyn_ops.cpp [hunk 1]
```diff
@@ -131,10 +132,44 @@ struct find_const_4in_slice
...
 };
 
+/**
+ * Simplify dimensions_of to a literal when the input arugment has a static shape
+ */
+struct find_static_dimensions_of
+{
+    auto matcher() const
+    {
+        return match::name("dimensions_of")(match::arg(0)(match::static_shape()));
```
[the remark this answers is not in the document]
Comment (umangyadav) [reply src/simplify_dyn_ops.cpp:142]: yeah i think would be good to expand `static_shape()` to check if the shape is `fixed` or not.

**Review by umangyadav — COMMENTED** [review]
Looks good. Perhaps would be good to expand on the `static_shape()` matcher. 
Or else the matcher can just match for `dimensions_of` and then do necessary checks inside `apply()`.

File: src/simplify_dyn_ops.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (CharlieL7) [reply src/simplify_dyn_ops.cpp:142]: We can do that in another matcher, however I doubt we should have a fully fixed dynamic shape to begin with. I cannot think of an operator that could do that. If the user supplies a fully fixed dynamic shape we parse it as a static shape.

[the remark this answers is not in the document]
Comment (umangyadav) [reply src/simplify_dyn_ops.cpp:142]: For this matcher however we only need to check if `start` to `end` axes have fixed dims or not.

File: src/simplify_dyn_ops.cpp [hunk 2]
```diff
@@ -131,10 +132,52 @@ struct find_const_4in_slice
...
+    {
+        auto ins                 = mr.result;
+        auto input               = ins->inputs().at(0);
+        auto dimensions_of_value = ins->get_operator().to_value();
+        auto start               = dimensions_of_value.at("start").to<std::size_t>();
+        auto end                 = dimensions_of_value.at("end").to<std::size_t>();
+        if(input->get_shape().dynamic())
+        {
+            // check if dynamic dimensions from start to end are fixed
+            auto dds = input->get_shape().dyn_dims();
```
Comment (umangyadav) [inline src/simplify_dyn_ops.cpp:153]: can you also add test that covers this condition ?

File: test/simplify_dyn_ops_test.cpp [hunk 3]
```diff
@@ -237,4 +237,32 @@ TEST_CASE(const_slice_4input)
...
+        m0.add_parameter("data", s);
+        migraphx::shape lit_shape{migraphx::shape::int64_type, {3}};
+        ;
+        std::vector<int64_t> lit_data = {2, 4, 4};
+        auto lit_ins                  = m0.add_literal(migraphx::literal{lit_shape, lit_data});
+        m0.add_return({lit_ins});
+    }
+
+    // dead_code_elimination will get rid of atan
+    migraphx::module m1;
```
Comment (pfultz2) [inline test/simplify_dyn_ops_test.cpp:254]: Can put this first? In the other tests the input module is first and the expected output is last.

**Review by umangyadav — APPROVED** [review]

**Comment by causten** [comment]
@CharlieL7  fix the format problem and this can merge

**Pull request merged**